src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,189 @@
+            return;
+        }
+
+        _isClosing = true;
+        lock (_openStreams)
+        {
+            _connectStream.Abort(exception, error);
+            foreach (var stream in _openStreams)
+            {
+                if (exception.InnerException is not null)
+                {
+                    stream.Value.AbortCore(new ConnectionAbortedException(exception.Message, exception.InnerException), error);
+                }
+                else
+                {
+                    stream.Value.AbortCore(new ConnectionAbortedException(exception.Message), error);
+                }
+            }
+        }
+
+        _openStreams.Clear();
+        _pendingStreams.Writer.Complete();
+    }
+
+    public async ValueTask<WebTransportStream> OpenUnidirectionalStreamAsync(CancellationToken cancellationToken)
+    {
+        if (_isClosing)
+        {
+            throw new ObjectDisposedException("WebTransport is closing the session");
+        }
+        // create the stream
+        var features = new FeatureCollection();
+        features.Set<IStreamDirectionFeature>(new DefaultStreamDirectionFeature(canRead: false, canWrite: true));
+        var connectionContext = await _connection._multiplexedContext.ConnectAsync(features, cancellationToken);
+        var streamContext = _connection.CreateHttpStreamContext(connectionContext);
+        var stream = new WebTransportStream(streamContext, WebTransportStreamType.Output);
+
+        // send the stream header
+        // https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#name-unidirectional-streams
+        await stream.WriteAsync(OutputStreamHeader, cancellationToken);
+        await stream.FlushAsync(cancellationToken);
+
+        return stream;
+    }
+
+    /// <summary>
+    /// Adds a new stream to the internal list of pending streams.
+    /// </summary>
+    /// <param name="stream">A reference to the new stream that is being added</param>
+    internal void AddStream(WebTransportStream stream)
+    {
+        if (_isClosing)
+        {
+            throw new ObjectDisposedException("WebTransport is closing the session");
+        }
+
+        if (!_pendingStreams.Writer.TryWrite(stream))
+        {
+            throw new Exception("Failed to add incoming stream to pending queue");
+        }
+    }
+
+    /// <summary>
+    /// Pops the first stream from the list of pending streams.
+    /// </summary>
+    /// <param name="cancellationToken">The cancellation token to abort waiting for a stream</param>
+    /// <returns>An instance of WebTransportStream that corresponds to the new stream to accept</returns>
+    public async ValueTask<WebTransportStream> AcceptStreamAsync(CancellationToken cancellationToken)
+    {
+        if (_isClosing)
+        {
+            throw new ObjectDisposedException("WebTransport is closing the session");
+        }
+
+        var stream = await _pendingStreams.Reader.ReadAsync(cancellationToken);
```

Comment:
The call throws the following exception:
`System.Threading.Channels.ChannelClosedException: 'The channel has been closed.'`
This then directly gets surfaced to the application. However, can this even happen? The only way the writer gets closed is if `_isClosing` is set to true, in which case the previous if will throw a more useful exception.

Reply:
It could actually be really common since an app could keep a pending accept call at all times.  If there were no pending streams when the accept started then the accept call could have already gotten past the _closed check and be waiting for the ReadAsync to finish. Then the channel is closed and the ReadAsync fails. We should catch that exception here and decide what to return to the user (null?).

Reply:
Ok